Draft: GSVA for RNA-seq (#396)
AlexsLemonade/refinebio-examples: 03-rnaseq/pathway-analysis_rnaseq_03_gsva.Rmd

  # Make an `Ensembl` column to store the row names
  tibble::rownames_to_column("Ensembl") %>%
  # Now let's join the rest of the expression data
  dplyr::inner_join(normalized_counts, by = c("Ensembl" = "ensembl_id"))
```

This `1:many mapping between keys and columns` message means that some Ensembl gene identifiers map to multiple Entrez IDs.
In this case, it's also possible that a Entrez ID will map to multiple Ensembl IDs.
For the purpose of performing GSVA later in this notebook, we keep only the first mapped IDs.
For more about how to explore this, take a look at our [microarray gene ID conversion example](https://alexslemonade.github.io/refinebio-examples/03-rnaseq/gene-id-annotation_rnaseq_01_ensembl.html).

Let's see a preview of `mapped_df`.

```{r}
head(mapped_df)
```

We will want to keep in mind that GSVA requires that data is in a matrix with the gene identifiers as row names.
In order to successfully turn our data frame into a matrix, we will need to ensure that we do not have any duplicate gene identifiers.

Let's check to see if we have any Entrez IDs that mapped to multiple Ensembl IDs.

```{r}
any(duplicated(mapped_df$entrez_id))
```

> Looks like we do have duplicated Entrez IDs.

Reviewer: I had a comment that's not outdated - how many? Probably worth talking about.

Author: ? What?

Reviewer: Should have been _now_ outdated. How many duplicated Entrez IDs are there? That is a good thing to talk about in the notebook itself.

Author: Gotcha.

**Review** 

#### Handling duplicate gene identifiers

As we mentioned earlier, we will not want any duplicate gene identifiers in our data frame when we convert it into a matrix in preparation for the GSVA steps later.
In other words, we will narrow down one Ensembl ID's data for each particular Entrez gene identifier.

See the [Gene Set Enrichment Analysis (GSEA) User guide](https://www.gsea-msigdb.org/gsea/doc/GSEAUserGuideFrame.html) for more information on other expression values that can be used, like the median expression value for example [@GSEA-user-guide].

First we need to calcuate the gene means, but we'll need to move our non-numeric variables (the gene ID columns) out of the way for that calculation. 

```{r}
# First let's determine the gene means
gene_means <- rowMeans(mapped_df %>% dplyr::select(-Ensembl, -entrez_id))

# Let's add this as a column in our `mapped_df`.
mapped_df <- mapped_df %>%

>   dplyr::mutate(gene_means) %>%

Reviewer: ```suggestion
  # Add gene_means as a column called gene_means
  dplyr::mutate(gene_means = gene_means) %>%
```

Maybe not both of these changes, but one of them for clarity.

      stringr::str_detect(refinebio_title, "_CV_") ~ "recovering"
    ),
    # It's easier for future items if this is already set up as a factor
    time_point = as.factor(time_point)
  ) %>%
  # We don't need the older version of the variable anymore
  dplyr::select(-refinebio_title)
```

Let's double check that our factor set up is right. 
We want `acute illness` to be the first level since it was the first time point collected. 

```{r}
levels(annot_df$time_point)
```

The `pheatmap::pheatmap()` will want the annotation data frame to have matching row names to the data we supply it (which is our `gsva_results`).

```{r}
annot_df <- annot_df %>%
  # pheatmap will want our sample names that match our data to
  tibble::column_to_rownames("refinebio_accession_code")
```

### Set up the heatmap itself

Great! We're all set. 
We can see that they are in a wide format with the gsva scores for each sample spread across a row associated with each pathway.

```{r}
heatmap <- pheatmap::pheatmap(gsva_results,
  annotation_col = annot_df, # Add metadata labels!
  show_colnames = FALSE, # Don't show sample labels
  fontsize_row = 6 # Shrink the labels a tad
)

# Print out heatmap here

> heatmap

Reviewer: I might add one line about something interesting in the heatmap and also it sounds like (from our 'in person' meeting) there are paired samples here that we also might want to _mention_ if not include as annotation.

Author: I tried including the labels for the paired samples, it just looked chaotic. But yeah, they deserve a mention.

We often suggest normalizing and transforming your data for various applications including for GSVA. 
We are going to use the `vst()` function from the `DESeq2` package to normalize and transform the data.
For more information about these transformation methods, [see here](https://alexslemonade.github.io/refinebio-examples/03-rnaseq/00-intro-to-rnaseq.html#deseq2-transformation-methods). 

```{r}
# Normalize and transform the data in the `DESeqDataSet` object using the `vst()`
# function from the `DESEq2` R package
dds_norm <- vst(dds)
```

At this point, if your data set has any outlier samples, you should look into removing them as they can affect your results. 
For this example data set, we will skip this step (there are no obvious outliers) and proceed.

Extract the normalized counts to a matrix and make it into a data frame so we can use with tidyverse functions later. 

```{r}
# Retrieve the normalized data from the `DESeqDataSet`
normalized_counts <- assay(dds_norm) %>%
  as.data.frame() %>% # Make into a data frame
  tibble::rownames_to_column("ensembl_id") # Make Gene IDs into their own column
```

**IDENTICAL TO MICROARRAY GSVA**

> ### Import Gene Sets

Reviewer: I have an idea for _variety_ that I'm not sure I'm super attached to but I'll put it out there - what if we went from `msigdbr` data frame to list format required fro GSVA? That would be a use/series of steps that was more generally applicable to organisms besides human.

Author: Worth a try!

Author: I don't hate it. See what you think: https://github.com/AlexsLemonade/refinebio-examples/pull/396/commits/9bafff4f44c3d70543a85db38548fc25be508fd8

Reviewer: I think it's good (although I will probably have thoughts about specific wording). Didn't add too many lines!

The function that we will use to run GSVA wants the gene sets to be in a list, rather than a tidy data frame.

> We are therefore going to read in the Hallmark collection of gene sets file directly from [MSigDB](https://www.gsea-msigdb.org/gsea/msigdb/index.jsp), rather than using the [`msigdbr`](https://cran.r-project.org/web/packages/msigdbr/index.html) package like we did in earlier notebooks [@Igor2020].

Reviewer: This will probably get changed in the other GSVA example when I #371, but there's not a notion of "earlier notebooks" in this context like there is in training. Sentiment expressed in https://github.com/AlexsLemonade/refinebio-examples/pull/359#discussion_r527883809 but at a different line.

Author: Oh yeah. I think this is copied directly from the microarray GSVA, but neither should have that.

  # Now let's join the rest of the expression data
  dplyr::inner_join(normalized_counts, by = c("Ensembl" = "ensembl_id"))
```

This `1:many mapping between keys and columns` message means that some Ensembl gene identifiers map to multiple Entrez IDs.
In this case, it's also possible that a Entrez ID will map to multiple Ensembl IDs.
For the purpose of performing GSVA later in this notebook, we keep only the first mapped IDs.
For more about how to explore this, take a look at our [microarray gene ID conversion example](https://alexslemonade.github.io/refinebio-examples/03-rnaseq/gene-id-annotation_rnaseq_01_ensembl.html).

Let's see a preview of `mapped_df`.

```{r}
head(mapped_df)
```

We will want to keep in mind that GSVA requires that data is in a matrix with the gene identifiers as row names.
In order to successfully turn our data frame into a matrix, we will need to ensure that we do not have any duplicate gene identifiers.

Let's check to see if we have any Entrez IDs that mapped to multiple Ensembl IDs.

```{r}
any(duplicated(mapped_df$entrez_id))
```

Looks like we do have duplicated Entrez IDs.
Let's find out which Entrez IDs have been duplicated.

> ```{r}

Author: We could delete this vector part, it's not as necessary now. In addition to 486 - 489

Author: I changed it to something more simple and straightforward.

> dup_entrez_ids <- mapped_df %>%

Reviewer: How many are there? Probably worth talking about.

  dplyr::filter(duplicated(entrez_id)) %>%
  dplyr::pull(entrez_id)
```

**Review** 

#### Handling duplicate gene identifiers

As we mentioned earlier, we will not want any duplicate gene identifiers in our data frame when we convert it into a matrix in preparation for the GSVA steps later.

> In other words, we will narrow down one Ensembl ID's data for a particular Entrez gene identifier instances for each sample or column in this case.

Reviewer: I don't think this is still true or maybe I'm not following. We'll also probably want to talk a little bit about the _why_ here, too. (We're in the draft stage, but wanted to let you know that there was a need for some rewriting coming up!)

Author: Yeah, there's a lot of the wording that needs polishing/fixing but I hadn't done that yet.

See the [Gene Set Enrichment Analysis (GSEA) User guide](https://www.gsea-msigdb.org/gsea/doc/GSEAUserGuideFrame.html) for more information on other expression values that can be used, like the median expression value for example [@GSEA-user-guide].

First we need to calcuate the gene means, but we'll need to move our non-numeric variables (the gene ID columns) out of the way. 

```{r}
# First let's determine the gene means
gene_means <- rowMeans(mapped_df %>% dplyr::select(-Ensembl, -entrez_id))

# Let's add this as a column in our `mapped_df`.
mapped_df <- mapped_df %>%
  dplyr::mutate(gene_means) %>%
  # Reorder the columns so `gene_means` column is upfront
  dplyr::select(Ensembl, entrez_id, gene_means, dplyr::everything())
```

Now we can filter out the duplicate gene identifiers using the mean expression value.
If we do an `dplyr::arrange()` by `gene_means` first, then this means the first row encountered will be the one with the higher gene mean and `dplyr::distinct()` keeps the first row of any duplicates it finds by default. 

```{r}
filtered_mapped_df <- mapped_df %>%

>   # Sort so that the highest absolute mean expression values are at the top

Reviewer: ```suggestion
  # Sort so that the highest mean expression values are at the top
```

The pathway-level score calculated is a way of asking how genes _within_ a gene set vary as compared to genes that are _outside_ of that gene set [@Malhotra2018].

The idea here is that we will get pathway-level scores for each sample that indicate if genes in a pathway vary concordantly in one direction (over-expressed or under-expressed relative to the overall population) [@Hanzelmann2013].

The output is a gene set by sample matrix of GSVA scores.

### Perform GSVA

Let's perform GSVA using the `gsva()` function.
See `?gsva` for more options.

```{r}
gsva_results <- gsva(
  filtered_mapped_matrix,
  hallmarks_list,
  method = "gsva",
  # Appropriate for our normalized counts data
  kcdf = "Gaussian",
  # Minimum gene set size
  min.sz = 15,
  # Maximum gene set size
  max.sz = 500,
  # Compute Gaussian-distributed scores
  mx.diff = TRUE,
  # Don't print out the progress bar
  verbose = FALSE
)
```

Note that the `gsva()` function documentation says we can use `kcdf = "Gaussian"` if we have expression values that are continuous such as log-CPMs, log-RPKMs or log-TPMs, but we would use `kcdf = "Poisson"` on integer counts.

> Our `vst()` transformed data is log, so `Gaussian` works for us. 

Reviewer: ```suggestion
Our `vst()` transformed data is on a log2-like scale, so `Gaussian` works for us. 
```
Or something that is more than "is log."

For this study, PBMCs were collected at two time points: during the patients' first, acute bronchiolitis visit (abbreviated "AV") and their recovery visit, (called post-convalescence and abbreviated "CV").

We can create a new variable, `time_point`, that states this info more clearly. 
This new `time_point` variable will have two labels: `acute illness` and `recovering` based on the `AV` or `CV` coding located in the `refinebio_title` string variable. 

```{r}
annot_df <- metadata %>%
  # We need the sample IDs and the main column that contains the metadata info
  dplyr::select(
    refinebio_accession_code,
    refinebio_title
  ) %>%
  # Create our `time_point` variable based on `refinebio_title`
  dplyr::mutate(
    time_point = dplyr::case_when(
      # Create our new variable based on refinebio_title containing AV/CV
      stringr::str_detect(refinebio_title, "_AV_") ~ "acute illness",
      stringr::str_detect(refinebio_title, "_CV_") ~ "recovering"
    ),
    # It's easier for future items if this is already set up as a factor
    time_point = as.factor(time_point)
  ) %>%
  # We don't need the older version of the variable anymore

>   dplyr::select(-refinebio_title)

Reviewer: Do we need these steps?

Author: Which part? The refinebio_title has unintelligible items like: `"AVB_111_CV_PBMC"` and there's not a metadata column that has this information separated out so we need to do some rearrangement. 

But any columns left in this data frame are automatically added to the plot by pheatmap, and I don't find that useful so that's why I'm dropping this column after we get the info from it. (thought about using a transmute but it ended up more confusing).

Reviewer: > But any columns left in this data frame are automatically added to the plot by pheatmap, and I don't find that useful so that's why I'm dropping this column after we get the info from it. (thought about using a transmute but it ended up more confusing).

Okay, makes sense. But the factor conversion?

Author: Oh factor conversion can probably be dropped now. 🎉

```

Let's double check that our factor set up is right. 
We want `acute illness` to be the first level since it was the first time point collected. 

```{r}
levels(annot_df$time_point)

> ```

Reviewer: Do we still care about this? Is it something to do with the heatmap legend?

Author: OH right, we don't care about factor levels now that we aren't doing DE.

The `pheatmap::pheatmap()` will want the annotation data frame to have matching row names to the data we supply it (which is our `gsva_results`).

```{r}
annot_df <- annot_df %>%
  # pheatmap will want our sample names that match our data to
  tibble::column_to_rownames("refinebio_accession_code")
```

### Set up the heatmap itself

> Great! We're all set. 

Reviewer: Do we need to check the order? I don't remember with `pheatmap()`.

Author: It goes by row name matching: 
pheatmap docs: 
> The rows in the data and in the annotation are matched using corresponding row names.

> We can see that they are in a wide format with the gsva scores for each sample spread across a row associated with each pathway.

Reviewer: ```suggestion
We can see that they are in a wide format with the GSVA scores for each sample spread across a row associated with each pathway.
```

Author: I think this also was a copy from Microarray, so I'll keep a list of little things that need updating there.

Reviewer: Can you stick them on #371? And I can feed two birds with one scone. (Birds shouldn't eat that much scone.)

```{r}

> heatmap <- pheatmap::pheatmap(gsva_results,

Reviewer: I _believe_ `heatmap()` is a built in graphics function, so would recommend changing this if that is the case.

Reviewer: Do we want to scale the rows here?

Author: I tried scaling. Seemed to make it look worse. Like the values were squashed and not in a good way.

  annotation_col = annot_df, # Add metadata labels!
  show_colnames = FALSE, # Don't show sample labels

>   fontsize_row = 6 # Shrink the labels a tad

Reviewer: ```suggestion
  fontsize_row = 6 # Shrink the pathway labels a tad
```

---
title: "Gene set variation analysis - RNA-seq"
author: "CCDL for ALSF"
date: "November 2020"
output:  
  html_notebook:
    toc: true
    toc_float: true
    number_sections: true
---

**IDENTICAL TO MICROARRAY GSVA** (except microarray -> rna-seq)

> # Purpose of this analysis

Author: I might not touch this until we see what happens for #371

Reviewer: That seems reasonable.

file.exists(metadata_file)
```

If the chunk above printed out `FALSE` to either of those tests, you won't be able to run this analysis _as is_ until those files are in the appropriate place.

If the concept of a "file path" is unfamiliar to you; we recommend taking a look at our [section about file paths](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#an-important-note-about-file-paths-and-Rmds). 

# Using a different refine.bio dataset with this analysis?

If you'd like to adapt an example analysis to use a different dataset from [refine.bio](https://www.refine.bio/), we recommend placing the files in the `data/` directory you created and changing the filenames and paths in the notebook to match these files (we've put comments to signify where you would need to change the code).
We suggest saving plots and results to `plots/` and `results/` directories, respectively, as these are automatically created by the notebook.
From here you can customize this analysis example to fit your own scientific questions and preferences. 

***

<!-- Do not delete this line --> <a name="analysis" style="padding-top:56px;margin-top:-56px;">&nbsp;</a>

> **IDENTICAL TO MICROARRRAY GSVA** (Except DESeq2 package added)

Author: Oh yup. I thought I added that but I apparently only did the `library()` add in the next chunk.

## GSVA - Microarray

Rather than ranking genes based on some gene-level statistic selected by the user, GSVA fits a model and ranks genes based on their expression level relative to the sample distribution.
The pathway-level score calculated is a way of asking how genes _within_ a gene set vary as compared to genes that are _outside_ of that gene set [@Malhotra2018].

The idea here is that we will get pathway-level scores for each sample that indicate if genes in a pathway vary concordantly in one direction (over-expressed or under-expressed relative to the overall population) [@Hanzelmann2013].

The output is a gene set by sample matrix of GSVA scores.

**STILL mostly IDENTICAL TO MICROARRAY GSVA**

### Perform GSVA

Let's perform GSVA using the `gsva()` function.
See `?gsva` for more options.

```{r}
gsva_results <- gsva(
  filtered_mapped_matrix,
  hallmarks_list,
  method = "gsva",
  # Appropriate for our normalized counts data
  kcdf = "Gaussian",
  # Minimum gene set size
  min.sz = 15,
  # Maximum gene set size
  max.sz = 500,
  # Compute Gaussian-distributed scores
  mx.diff = TRUE,
  # Don't print out the progress bar
  verbose = FALSE
)
```

> Here we are using `kcdf = "Poisson"` because we are using normalized and transformed counts. 

Author: Will add that back in.
